oximeter/db/src/oxql/ast/grammar.rs

```diff
@@ -327,66 +328,132 @@ peg::parser! {
             lit
         }
 
+        /// Parse a logical operator.
         pub(super) rule logical_op_impl() -> LogicalOp
-            = "||" { LogicalOp::Or} / "&&" { LogicalOp::And }
-
-        // Parse a filter item, which is one logical expression used in a filter
-        // operation.
-        #[cache_left_rec]
-        pub(super) rule filter_item() -> Filter = precedence! {
-            // Note: We need to separate the logical operations into different
-            // levels of precedence.
-            left:(@) _? "||" _? right:@ {
-                Filter::Expr(FilterExpr {
-                    left: Box::new(left),
-                    op: LogicalOp::Or,
-                    right: Box::new(right),
-                })
-            }
-            --
-            left:(@) _? "&&" _? right:@ {
-                Filter::Expr(FilterExpr {
-                    left: Box::new(left),
-                    op: LogicalOp::And,
-                    right: Box::new(right),
-                })
-            }
-            --
-            a:filter_atom() { Filter::Atom(a) }
-            "(" e:filter_item() ")" { e }
-        }
-
-        rule filter_atom() -> FilterAtom
-            = atom:(negated_filter_atom() / unnegated_filter_atom())
+            = "||" { LogicalOp::Or}
+            / "&&" { LogicalOp::And }
+            / "^" { LogicalOp::Xor }
+
+
+        // NOTES:
+        //
+        // The rules below are all used to parse a filtering expression. This
+        // turns out to be surprisingly complicated to express succinctly in
```

Comment:
Oh man, I'm with you. I very nearly ripped this all out and wrote a manual recursive-descent parser. In the end, I think it's _okay_ where it is, but I'm certainly on board with moving in that direction long term. For future reference, while developing this I found a pretty nice crate called [`logos`](https://docs.rs/logos/latest/logos/), which is a **lexer generator**. Lexing is more tedious, so it might be helpful to generate the lexer, and write a manual parser. Just a thought.